graphql_compiler/deserialization.py

```diff
@@ -0,0 +1,253 @@
+
+from graphql import (
+    GraphQLBoolean,
+    GraphQLFloat,
+    GraphQLID,
+    GraphQLInt,
+    GraphQLList,
+    GraphQLScalarType,
+    GraphQLString,
+)
+
+from graphql_compiler import GraphQLInvalidArgumentError
+from graphql_compiler.compiler.helpers import strip_non_null_from_type
+from graphql_compiler.global_utils import is_same_type
+from graphql_compiler.typedefs import QueryArgumentGraphQLType
+
+from .global_utils import assert_set_equality
+from .schema import SUPPORTED_SCALAR_TYPES, GraphQLDate, GraphQLDateTime, GraphQLDecimal
+
+
+_ALLOWED_JSON_SCALAR_TYPES: Mapping[str, Tuple[Type, ...]] = MappingProxyType(
+    {
+        GraphQLDate.name: (str,),
+        GraphQLDateTime.name: (str,),
+        GraphQLFloat.name: (str, float, int),
+        GraphQLDecimal.name: (str, float, int),
+        GraphQLInt.name: (int, str),
+        GraphQLString.name: (str,),
+        GraphQLBoolean.name: (bool, int, str),
+        GraphQLID.name: (int, str,),
+    }
+)
+assert_set_equality(
+    set(_ALLOWED_JSON_SCALAR_TYPES.keys()),
+    {graphql_type.name for graphql_type in SUPPORTED_SCALAR_TYPES},
+)
+
+
+def _custom_boolean_deserialization(value: Any) -> bool:
+    """Deserialize a boolean, allowing for common string or int representations."""
+    true_values = [1, "1", "true", "True", True]
+    false_values = [0, "0", "false", "False", False]
+    if value in true_values:
+        return True
+    elif value in false_values:
+        return False
+    else:
+        raise ValueError(
+            f"Received unexpected GraphQLBoolean value {value} ({type(value)}). Expected one "
+            f"of the following {true_values + false_values}."
+        )
+
+
+_CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS: Mapping[str, Callable[[Any], Any]] = MappingProxyType(
+    {
+        # Bypass the GraphQLFloat parser and allow strings as input. The JSON spec allows only
+        # for 64-bit floating point numbers, so large floats might have to be represented as
+        # strings.
+        GraphQLFloat.name: float,
+        # Bypass the GraphQLInt parser and allow long ints and strings as input. The JSON spec
+        # allows only for 64-bit floating point numbers, so large ints might have to be
+        # represented as strings.
+        GraphQLInt.name: int,
+        # Bypass the GraphQLBoolean parser and allow some strings and ints as input.
+        GraphQLBoolean.name: _custom_boolean_deserialization,
+    }
+)
+
+_JSON_TYPES_AND_DESERIALIZATION_FUNCTIONS: Mapping[
+    str, Tuple[Tuple[Type, ...], Callable[[Any], Any]]
+] = MappingProxyType(
+    {
+        scalar_type.name: (
+            _ALLOWED_JSON_SCALAR_TYPES[scalar_type.name],
+            _CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS.get(scalar_type.name, scalar_type.parse_value),
+        )
+        for scalar_type in SUPPORTED_SCALAR_TYPES
+    }
+)
+
+
+def convert_scalar_value_to_graphql_type(expected_type: GraphQLScalarType, value: Any) -> Any:
```

Comment:
How about `deserialize_scalar_value` instead?  It is shorter and `convert_scalar_value_to_graphql_type` is not really accurate since this function does not return a GraphQL type.